src/tsdb.c

```diff
@@ -201,23 +220,109 @@ size_t SeriesGetNumSamples(const Series *series) {
+            // TODO: check if new sample before first sample
+            RedisModule_DictIteratorStop(dictIter);
+            assert(0);
+            return TSDB_ERROR;
+        }
+    }
+    Sample sample = {.timestamp = timestamp, .value = value };
+    AddCtx aCtx = { .sz = 0,
+                    .inChunk = chunk,
+                    .sample = sample,
+                    .maxSamples = series->maxSamplesPerChunk,
+                    .type =  UPSERT_ADD // TODO: on-conflict param
+                    };
+    // TODO
+    int rv = series->funcs->UpsertSample(&aCtx);
+    if (rv == REDISMODULE_OK) {
+        series->totalSamples += aCtx.sz;
+
+        // reindex if first timestamp changed
+        if (aCtx.reindex == true) {
+            if (chunkKey) {
+                RedisModule_DictDelC(series->chunks, chunkKey, sizeof(timestamp_t), NULL);
+            } else {
+                dictOperator(series->chunks, NULL, lastChunkFirstTS, DICT_OP_DEL);
+            }
+            dictOperator(series->chunks, chunk, timestamp, DICT_OP_SET);
+        }
+        upsertHandleRules(series, &aCtx);
+    }
+    return rv;
+}
+
 int SeriesAddSample(Series *series, api_timestamp_t timestamp, double value) {
-    timestamp_t rax_key;
-    if (timestamp <= series->lastTimestamp && series->lastTimestamp != 0) {
+    if (!(series->options & SERIES_OPT_OUT_OF_ORDER)) {
+        if (timestamp < series->lastTimestamp && series->lastTimestamp != 0) {
+            return TSDB_ERR_TIMESTAMP_TOO_OLD;
+        } else if (timestamp == series->lastTimestamp && timestamp != 0) {
+            return TSDB_ERR_TIMESTAMP_OCCUPIED;
+        }
+    // TODO: ensure cover for out of retention insertion.
+    } else if (series->retentionTime && 
+               timestamp < timestamp - series->retentionTime) {
+        // TODO: downsample window is partially trimmed
         return TSDB_ERR_TIMESTAMP_TOO_OLD;
-    } else if (timestamp == series->lastTimestamp && timestamp != 0) {
-        return TSDB_ERR_TIMESTAMP_OCCUPIED;
     }
+
+    // backfilling or update
+    if (timestamp <= series->lastTimestamp && series->totalSamples != 0) {
+        return SeriesUpsertSample(series, timestamp, value);
+    }
+
     Sample sample = {.timestamp = timestamp, .value = value};
-    int ret = series->funcs->AddSample(series->lastChunk, &sample);
+    ChunkResult ret = series->funcs->AddSample(series->lastChunk, &sample);
+
     if (ret == CR_END) {
         // When a new chunk is created trim the series
         SeriesTrim(series);
 
         Chunk_t *newChunk = series->funcs->NewChunk(series->maxSamplesPerChunk);
-        seriesEncodeTimestamp(&rax_key, timestamp);
-        RedisModule_DictSetC(series->chunks, &rax_key, sizeof(rax_key), (void*)newChunk);
-        series->funcs->AddSample(newChunk, &sample);
+        dictOperator(series->chunks, newChunk, timestamp, DICT_OP_SET);
+        ret = series->funcs->AddSample(newChunk, &sample);
+        assert(ret == CR_OK);
```

Comment:
we should assert here, check the value, clean what needed to be cleaned (im not sure there's any cleanup needed) and return an error.

Reply:
We have just allocated a new chunk, this should not fail

Reply:
So why we need to assert it? we don't do that anywhere and I don't see why we should do it now.
`malloc` rarely fails if at all in modern POSIX boxes with memory over commit.